content/material/2020-01-17-intermediate-dplyr.Rmd

```diff
@@ -0,0 +1,389 @@
+diamonds_reshaped
+```
+
+
+### Handle dates with lubridate
+
+Often enough, data are collected on the same subjects (in this case storm names) across time, this implied the use of dates. However, handling dates can become a bit of a headache if R does not recognize the format. 
+The package lubridate helps specifically with this problem and allows to format dates in a way that can be unequivocally recognized by R.
+
+
+Although lubridate is part of the tidyverse (and we do not need to install it separately), we need to call it separately because it does not get loaded when we load the tidyverse library.
+
+```{r}
+library(lubridate)
+```
+
+The best way to input dates (if you are the one doing the data input), is to create three separate columns for day, month, year.
+
+Then we can use the function `ymd()` from the package lubridate in combination with the function `paste()` to create a date variable
+
+```{r}
+storms_with_date <- storms %>% 
+  mutate(date = ymd(paste(year, month, day, sep = "-") )) 
+```
+
+if you query the new dataframe using the function `View()` you can see that the new variable date is specifically formatted as a Date data type.
+
+```{r}
+str(storms_with_date)
+```
+
+# Extra: Tidying messing data
+
+So far, we have only reshaped "clean" datasets, but at major usecase for the `pivot_**` functions is tidying up messy data. In the example below, we will tidy a made up (but very realistic) table.
+
+```{r}
+blood_pressure <- tibble(
+  id = 1:3, 
+  bp_Sys_week1 = rnorm(3, 130, 5), bp_Dia_week1 = rnorm(3, 90, 5), weight_week1 = rnorm(3, 80, 10),
+  bp_Sys_week2 = rnorm(3, 130, 5), bp_Dia_week2 = rnorm(3, 90, 5), weight_week2 = weight_week1 + rnorm(3),
+  bp_Sys_week3 = rnorm(3, 130, 5), bp_Dia_week3 = rnorm(3, 90, 5), weight_week3 = weight_week2 + rnorm(3)
+)
+
+blood_pressure
+
+```
+
+The main problem in this table is that the variable names also inclue a value (week number). We want to pivot this wide dataframe into a longer one, with each row containg an id, a week number, a systolic and a diastolic blood pressure and a weigth. For this we use `pivot_longer()`.
+
+We do not want to include `id` in the reshape, so we select every column except this with `-id` as the first argument. We then seperate the variable names (e.g. 'bp_Dia_week1') into the actual variable name ( 'bp_Dia') and the value of the week variable ('1'). This is donw by setting the `names_sep` parameter to the string that seperates the variable name from the week value (here '_week'). Because we split each variable name into two, we must set `names_to` for both. We do this with the character vector `c('.value', 'week')`. The special value `'.value'` indicates that the corresponding substring of the original variable name will itself become a variable name for the values in the new dataframe (This makes a lot more sense when you look at the output of the function). The parameter `names_ptypes` convertes the variable `week` from a character to an integer.
```

Comment:
I'm not too proud of this formulations, but I think it is a very important use case for the pivot functions. Generally pivoting is an abstraction that is very hard to explain, but quite quickly becomes second nature.